infra/Pos/Slotting/Types.hs

```diff
@@ -22,15 +37,96 @@ data EpochSlottingData = EpochSlottingData
 
 instance NFData EpochSlottingData
 
+
 -- | Data necessary for slotting to work which is basically part of GState.
-data SlottingData = SlottingData
-    { sdPenult      :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for penult (i. e. 'sdPenultEpoch') epoch.
-    , sdLast        :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for last epoch.
-    -- Last means 'sdPenultEpoch + 1'.
-    , sdPenultEpoch :: !EpochIndex
-    -- ^ Penult epoch for which 'EpochSlottingData' is known.
-    } deriving (Eq, Show, Generic)
+-- External code can use functions like getLastEpochIndex or getLastEpochSlottingData and 
+-- not worry about cases where it doesn't exist (this module should be responsible for it).
+-- Note that it's important to use error rather than default values like 0, because 
+-- such cases indicate invariants violation and shouldn't be hidden behind default values.
+type PenultSlottingData = EpochSlottingData
+
+newtype SlottingData = SlottingData 
+    { getSlottingDataMap :: Map EpochIndex EpochSlottingData 
+    } deriving (Eq, Show, Generic, Monoid)
 
 instance NFData SlottingData
+
+-- | Unsafe constructor that can lead to invalid state!
+createSlottingDataUnsafe :: Map EpochIndex EpochSlottingData -> SlottingData
+createSlottingDataUnsafe = SlottingData
+
+-- | Restricted constructor function for the (initial) creation of @SlottingData@.
+-- Optional: Wrap @EpochSlottingData@ into @PenultSlottingData@ using newtype.
+createInitSlottingData 
+    :: PenultSlottingData 
+    -> EpochSlottingData 
+    -> SlottingData
+createInitSlottingData psd esd = SlottingData validInitialSlottingData
```

Comment:
Again, I think we don't need penult epoch slotting data, as I wrote earlier. Probably I had to think earlier, sorry. It doesn't seem to give any benefit anymore and adds extra complication.
So previously we had effectively a map with exactly two elements and now we are going to have a map with at least 1 element.